File: Swimat/Prefs.m
#import "Prefs.h"

@implementation Prefs

NSString * const TAG_AUTO_SAVE = @"auto_save";
NSString * const TAG_AUTO_ON_BUILD = @"auto_on_build";
NSString * const TAG_INDENT_EMPTY_LINE = @"IndentEmptyLine";
NSString * const TAG_INDENT = @"indent";
NSString * const INDENT_TAB = @"Tab Indent";
NSString * const INDENT_SPACE2 = @"2 Space Indent";
NSString * const INDENT_SPACE4 = @"4 Space Indent";
static NSString *const BREAK_BEFORE_OPENING_BRACE = @"BreakBeforeOpeningBraceRule";
static NSString *const BREAK_BEFORE_OPENING_BRACE_ITEM_TITLE_IGNORE = @"Ignore";
static NSString *const BREAK_BEFORE_OPENING_BRACE_ITEM_TITLE_REMOVE = @"Remove";
static NSString *const BREAK_BEFORE_OPENING_BRACE_ITEM_TITLE_FORCE = @"Force";

+(void) setIndent:(NSString *)value {
	NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
	[prefs setObject:value forKey:TAG_INDENT];
	[prefs synchronize];
}

+(NSString *) getIndent {
	NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
	NSString *value = [prefs stringForKey:TAG_INDENT];
	if (value == nil) {
		value = INDENT_TAB;
	}
	return value;
}

+(NSArray *) getIndentArray {
	return @[INDENT_TAB, INDENT_SPACE2, INDENT_SPACE4];
}

+(NSString *) getIndentString {
	NSString *tag = [self getIndent];
	if ([tag isEqualToString:INDENT_TAB]) {
		return @"\t";
	} else if ([tag isEqualToString:INDENT_SPACE2]) {
		return @"  ";
	} else if ([tag isEqualToString:INDENT_SPACE4]) {
		return @"    ";
	}
	return @"\t";
}

+(void) setAutoFormat:(bool) format {
	NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
	[prefs setBool:format forKey:TAG_AUTO_SAVE];
	[prefs synchronize];
}

+(bool) isAutoFormat {
	NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
	return [prefs boolForKey:TAG_AUTO_SAVE];
}

+(void) setFormatOnBuild:(bool) format {
	NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
	[prefs setBool:format forKey:TAG_AUTO_ON_BUILD];
	[prefs synchronize];
}

+(bool) isFormatOnBuild {
	NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
	return [prefs boolForKey:TAG_AUTO_ON_BUILD];
}

+(void) setIndentEmptyLine:(bool) format {
	NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
	[prefs setBool:format forKey:TAG_INDENT_EMPTY_LINE];
	[prefs synchronize];
}

+(bool) isIndentEmptyLine {
	NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
	return [prefs boolForKey:TAG_INDENT_EMPTY_LINE];
}

#pragma mark - Formatting rule: Break before opening brace

+ (void)setBreakBeforeOpeningBraceRule:(SWMBreakBeforeOpeningBraceRule)rule {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    [defaults setInteger:rule forKey:BREAK_BEFORE_OPENING_BRACE];
    [defaults synchronize];
}

+ (SWMBreakBeforeOpeningBraceRule)breakBeforeOpeningBraceRule {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    id setting = [defaults objectForKey:BREAK_BEFORE_OPENING_BRACE];
    if (setting == nil) {
        // Initialize setting with default rule
        [defaults setInteger:SWMBreakBeforeOpeningBraceRuleIgnore forKey:BREAK_BEFORE_OPENING_BRACE];
        [defaults synchronize];
    }
    
    return [defaults integerForKey:BREAK_BEFORE_OPENING_BRACE];
}

+ (NSString *)menuItemTitleForBreakBeforeOpeningBraceRule:(SWMBreakBeforeOpeningBraceRule)rule {
    switch (rule) {
        case SWMBreakBeforeOpeningBraceRuleIgnore: {
            return NSLocalizedString(BREAK_BEFORE_OPENING_BRACE_ITEM_TITLE_IGNORE, nil);
        }
            
        case SWMBreakBeforeOpeningBraceRuleRemove: {
            return NSLocalizedString(BREAK_BEFORE_OPENING_BRACE_ITEM_TITLE_REMOVE, nil);
        }
            
        case SWMBreakBeforeOpeningBraceRuleForce: {
            return NSLocalizedString(BREAK_BEFORE_OPENING_BRACE_ITEM_TITLE_FORCE, nil);
        }
    }
}

+ (NSString *)enabledBreakBeforeOpeningBraceRuleMenuItemTitle {
    SWMBreakBeforeOpeningBraceRule enabledRule = [self breakBeforeOpeningBraceRule];
    return [self menuItemTitleForBreakBeforeOpeningBraceRule:enabledRule];
}

@end

File: Swimat/Prefs.swift
import Foundation

class Prefs {

	private static let Indent = "indent"
	private static let IndentDefault = "\t"
	static let IndentTitle = ["Tab Indent", "2 Space Indent", "3 Space Indent", "4 Space Indent"]
	static let IndentTable = ["Tab Indent": "\t", "2 Space Indent": "  ", "3 Space Indent": "   ", "4 Space Indent": "    "]
	private static let SaveTrigger = "save_trigger"
	private static let IndentEmptyLine = "indent_emptyline"
	private static let SmartLine = "smart_line"

	static func setIndent(indent: String) {
		setString(Indent, value: indent)
	}

	static func getIndent() -> String {
		return getString(Indent, defValue: IndentDefault)
	}

	static func saveTrigger(trigger: Bool) {
		setBool(SaveTrigger, value: trigger)
	}

	static func isSaveTrigger() -> Bool {
		return getBool(SaveTrigger, defValue: false)
	}

	static func indentEmptyLine(trigger: Bool) {
		setBool(IndentEmptyLine, value: trigger)
	}

	static func isIndentEmptyLine() -> Bool {
		return getBool(IndentEmptyLine, defValue: false)
	}

	static func setSmartLine(value: Bool) {
		setBool(SmartLine, value: value)
	}

	static func isSmartLine() -> Bool {
		return getBool(SmartLine, defValue: false)
	}

}

extension Prefs { // MARK: base function

	static func getBool(tag: String, defValue: Bool) -> Bool {
		let userDefault = NSUserDefaults.standardUserDefaults()
		return userDefault.boolForKey(tag) ?? defValue
	}

	static func setBool(tag: String, value: Bool) {
		let userDefault = NSUserDefaults.standardUserDefaults()
		userDefault.setBool(value, forKey: tag)
	}

	static func getString(tag: String, defValue: String) -> String {
		let userDefault = NSUserDefaults.standardUserDefaults()
		return userDefault.stringForKey(tag) ?? defValue
	}

	static func setString(tag: String, value: String) {
		let userDefault = NSUserDefaults.standardUserDefaults()
		userDefault.setObject(value, forKey: tag)
	}

}
